### ISSUE TITLE
Slow backward() speed when using torch.nn.parallel.DistributedDataParallel

### ISSUE THREAD

**Issue by X-Lai** [yassouali/CCT#8]
Thanks a lot for your great work and clean repo.  @yassouali 

I have run your code on two gpus, and got at best 70.9 mIoU under your default config. However, it took me 28 hours to train on 2 titan-xp gpus, which is somewhat slow. Also, taking the sync batchnorm into consideration, I want to equip your code with pytorch official DistributedDataParallel and its syncBN to accelerate training and  increase accuracy. But I found that it's even slower than the normal DataParallel, and the best performance is only 69.9 at best. After checking, with DistributedDataParallel the 'loss.backward()' operation consumes ~1.3s each step while with normal DataParallel it only consumes ~0.4s.  And the gradients slightly differ from the normal DataParallel.

Do you make any optimization on your code with the 'backward()' operation? Have you tried to use DistributedDataParallel? How to apply syncBN to your code? Is there any way to accelerate the training? And Would you mind giving me some advice on my problem?

Looking forward to your reply. Thanks a lot.

**Comment by yassouali** [comment]
Hi @X-Lai 

Thank your for you interest in our work.

Sadly, I did not test with DistributedDataParallel nor syncBN.
In order to reduce the backward time without losing too much performance, you can try reducing the number of aux. decoders K, I think K=2 works quite well, you can also remove some decoders that are more computationally expensive than other, for example, I-VAT and F-Noise are more expensive than the rest.

**Comment by X-Lai** [comment]
Thanks a lot for your reply. Now the pytorch DistributedDataParallel works well in RTX 2080ti. I guess that Titan xp doesn't support much optimization for DistributedDataParallel, which causes the slow speed of backward.
